File: searching_and_sorting/split_arr_largest_sum.js

let arr = [7, 2, 5, 10, 8];
let k = 2;

function subArr(arr, mid) {
    let sub = 1;
    let sum = 0;
    for (let i = 0; i < arr.length; i++) {
        if (sum + arr[i] <= mid) {
            sum += arr[i];
        } else {
            sub++;
            sum = arr[i];
        }
    }
    return sub;
}

function splitArray(arr, k) {
    let n = arr.length;

    if (k > n) {
        return -1;
    }

    let low = Math.max(...arr);
    let high = arr.reduce((a, b) => a + b, 0);

    while (low <= high) {
        let mid = Math.floor((low + high) / 2);
        if (subArr(arr, mid) > k) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    return low;

}

console.log(splitArray(arr, k));

File: searching_and_sorting/split_arr_largest_sum.py
class Solution(object):
    def subArr(self, arr,mid):
        sub = 1
        sum = 0
        for i in range(0,len(arr)):
            if(sum+arr[i]<=mid):
                sum+=arr[i]
            else:
                sub=sub+1
                sum=arr[i]
        return sub
    
    
    def splitArray(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        n = len(nums)
        
        if(k>n):
            return -1
        low = max(nums)
        high = sum(nums)

        while (low<=high):
            mid = (low+high)/2
            if(self.subArr(nums,mid)>k):
                low=mid+1
            else:
                high=mid-1
        return low